test/generators/gen_link/gen_link.ml

```diff
@@ -0,0 +1,235 @@
+              Atom "odoc";
+              Atom "compile";
+              Atom "--pkg";
+              Atom "test";
+              Atom "-o";
+              Atom "%{target}";
+              Atom "%{deps}";
+            ];
+        ];
+    ]
+
+let odocl_target_rule path =
+  let open Fpath in
+  let basename = basename (rem_ext path) in
+  List
+    [
+      Atom "rule";
+      List [ Atom "target"; Atom (basename ^ ".odocl") ];
+      List [ Atom "deps"; Atom (basename ^ ".odoc") ];
+      List
+        [
+          Atom "action";
+          List
+            [
+              Atom "run";
+              Atom "odoc";
+              Atom "link";
+              Atom "-o";
+              Atom "%{target}";
+              Atom "%{deps}";
+            ];
+        ];
+    ]
+
+let file_rule path ext =
+  let open Printf in
+  printf "%s\n\n" (to_string_hum (cu_target_rule path ext));
+  printf "%s\n\n" (to_string_hum (odoc_target_rule path ext));
+  printf "%s\n\n" (to_string_hum (odocl_target_rule path))
+
+let mld_odoc_target_rule path =
+  let open Fpath in
+  let basename = basename (rem_ext path) in
+  List
+    [
+      Atom "rule";
+      List [ Atom "target"; Atom ("page-" ^ basename ^ ".odoc") ];
+      List [ Atom "deps"; Atom (to_string path) ];
+      List
+        [
+          Atom "action";
+          List
+            [
+              Atom "run";
+              Atom "odoc";
+              Atom "compile";
+              Atom "--pkg";
+              Atom "test";
+              Atom "-o";
+              Atom "%{target}";
+              Atom "%{deps}";
+            ];
+        ];
+    ]
+
+let mld_odocl_target_rule path =
+  let open Fpath in
+  let basename = basename (rem_ext path) in
+  List
+    [
+      Atom "rule";
+      List [ Atom "target"; Atom ("page-" ^ basename ^ ".odocl") ];
+      List [ Atom "deps"; Atom ("page-" ^ basename ^ ".odoc") ];
+      List
+        [
+          Atom "action";
+          List
+            [
+              Atom "run";
+              Atom "odoc";
+              Atom "link";
+              Atom "-o";
+              Atom "%{target}";
+              Atom "%{deps}";
+            ];
+        ];
+    ]
+
+let mld_file_rule path =
+  let open Printf in
+  printf "%s\n\n" (to_string_hum (mld_odoc_target_rule path));
+  printf "%s\n\n" (to_string_hum (mld_odocl_target_rule path))
+
+let die s =
+  prerr_endline s;
+  exit 1
+
+let path' () f = Filename.quote (Fpath.to_string f)
+
+let ext' () f = Filename.quote (Fpath.get_ext f)
+
+let gen_rule_for_source_file path =
+  let ext = Fpath.get_ext path in
+  match ext with
+  | ".ml" -> file_rule path ".cmt"
+  | ".mli" -> file_rule path ".cmti"
+  | ".mld" -> mld_file_rule path
+  | _ ->
+      die
+        (sprintf
+           "Don't know what to do with %a because of unrecognized %a extension."
+           path' path ext' path)
+
+let odocls paths =
+  paths
+  |> List.map (fun p ->
+         let open Fpath in
+         let path = relativize ~root:(v "html") p in
```

Comment:
You can put `v "html"` into a global variable and use it in `gen_rule_html` so the `"html"` constant isn't repeated.